### ISSUE TITLE
Logistic and Storage Center

### ISSUE THREAD

**Issue by TripleNipplFish** [satisfactory-factories/application#46]
Hello,

great work, love the concept. 

Would be cool to have the possibility to make storage and logistic centers where you just can import and export thinks without producing something.

**Comment by Maelstromeous** [comment]
This unfortunately breaks the concept of the site, it's kind of a use case that wasn't really designed for it. The primary reason for this is that storage centers are variable, they contain a lot of products that are ultimately fed by other factories and something has to produce the items that go into it. You loose an element of control of importing rates because the storage bursts it's output due to it's large buffers.

This tool isn't really designed for that concept, unfortunately. I appriciate this isn't the answer you really want, but it would be a substantial code change to support invisible "factories" that contain no products. It basically breaks the accounting math of the site.

In my personal opinion, with the advent of Dimensional Storage, there's no real need to have a centralized storage area any more. You can have up to 5 stacks of any item in the game in the DS, which for me nullifies the need for a physical storage area. It also hides issues with supply and demand by adding a buffer, which is inefficent and adds unneeded complexity when really you should deliver goods directly.

You could always still use your storage area, ultimately the factory producing the items is still delivering somewhere, it's just taking a long time to do so via a proxy.

I'll leave this issue open for a time in case someone has other ideas.

**Comment by TripleNipplFish** [comment]
Thanks for the response.

I’ve reviewed the code a bit. Unfortunately, I’m not a TypeScript or JavaScript Dev at all. For my needs, it seems like imports would need to be included as part of the calculated surplus. However, I assume this would require changes to the factory interface. I’m not asking you to implement this, but if I were to create an implementation where inputs could exceed production usage and be monitored in the export tab, would you consider merging that pull request?

**Comment by Maelstromeous** [comment]
@TripleNipplFish  when you say exceed production usage do you mean you have an import that's purposely higher than the requirements? As that is already possible now, it'll just show Trim next to it.

**Comment by samsmithnz** [comment]
I wonder if part of the answer here to just create a factory called 'storage', that imports in excess production? The only issue is that you need to add the product to the central storage so that you can select it for import, and then remove the product after (unless after adding the product - it first looks to see what is being imported before deciding what dependencies it needs?)

![Image](https://github.com/user-attachments/assets/b644ff10-d5dd-48f4-8e13-f3aca9f5a3e6)
![Image](https://github.com/user-attachments/assets/4837c745-3b34-4958-afd0-ff9f1f4ea650)

**Comment by TripleNipplFish** [comment]
@Maelstromeous Yes, that is what i mean. 

But the Import overflow isn't shown in the Export tab or am i doing something wrong? The goal is to be able to export from the logistic center.

**Comment by TripleNipplFish** [comment]
Realized, you just need the product with 0 up in products

**Comment by Maelstromeous** [comment]
@samsmithnz paradoxically this may be a bug, it shouldn't be possible to add an import without a product. We could of course remove that ability, but then the filtering of the imports will go away, leading the user to have to pick from a large list of factories.

Or, we redesign the import picker. I wasn't overly happy with it in the first place, there's bugs with it on how it picks it's choices for candidates as well. It may be better we redesign it, and present a better UI for it where factories that match the user's products are highlighted, then free-form for any factory thereafter.

**Comment by Maelstromeous** [comment]
I'm going to reopen this until we have a proper decision on how we want to handle this

**Comment by TripleNipplFish** [comment]
Fair, will also try to find a nice solution.

**Comment by samsmithnz** [comment]
I think if you could select a product and allow that to be filled by imports it would work? Does that make sense?

**Comment by TripleNipplFish** [comment]
I guess you could make a button at Imports to bypass the filter in constructDependencies ?

**Comment by 450Volts** [comment]
Just adding my 2 cents that supporting some kind of "storage" fed by unallocated factory surplus which can be drawn from would be immensely relevant to my playstyle since I prefer to use central distribution centers. This allows me worry less about balancing of dedicated production lines (that draw from specific upstream factories) and to focus on total production capacity of the region. E.g., When I'm considering expanding a modular frame factory, I don't really care what factory the iron rods are coming from, so far as there is sufficient surplus iron rods being produce I can draw from.

**Comment by samsmithnz** [comment]
Revisiting this after a discord conversation yesterday. A lot of the model restricts imports to relevant factories. For example, if the factory is building computers, the imports will only show factories that are exporting plastic, circuits, or cables. 
```mermaid
graph TD
    OilFactory --"100 Plastic"--> ComputerFactory
    OilFactory --"20 Plastic"--> Surplus
```

We are wondering about adding a 'distribution' or 'storage' checkbox, to a factory, which would essentially remove that requirement, and allow you to import any surplus products from any factory. There would need to be a clear description with the checkbox. 

The challenge then, is that this 'distribution' factory would now show in other factories as a valid import. Again, the example of the computer, if we put plastic here, the computer factory will draw from it again. 
```mermaid
graph TD
    OilFactory --"100 Plastic"--> ComputerFactory
    OilFactory --"20 Plastic"--> Distribution/Storage
    Distribution/Storage --"20 Plastic"--> ComputerFactory
```

Therefore, I think we should also consider the storage aspect of this - where you can mark a particular production line as going to 'storage' and hence, is not available as a surplus. 
- Now the excess part would not be available for export. 
- This would solve the current trim issue for end product production (e.g. a space elevator factory that produces smart plates is currently considered surplus, because it's not being used by anything - even though in the game it's going to the space elevator). 
```mermaid
graph TD
    OilFactory --"100 Plastic"--> ComputerFactory
    OilFactory --"20 Plastic"--> Distribution/Storage
```

**Comment by 450Volts** [comment]
@samsmithnz Thanks for the examples. However, there are a few aspects that seem confusing or unexpectedly complicated to me:

> We are wondering about adding a 'distribution' or 'storage' marker (perhaps a checkbox?), to a factory, which would essentially remove that requirement, and allow you to import any surplus products from any factory. 

I agree that we are looking for the best way to support less-restricted import to factories.

> There would need to be a clear description with this marker, as this could be confusing to some players._

Why is this so? It feels intuitive to me that any produced part could be imported to any factory at a chosen rate, just like players can in run a belt of any part to any factory in game. I feel the tool's logic is well positioned to communicate the expectation that every factory can import any part the user decides (so far as that part is produced anywhere else in the plan). Of course, most imported parts may be used to produce downstream parts directly, however (as we're discussing) this not necessary. Unifying the ability to freely import parts to any factory avoids the need to set separate rules for "normal" factories and "Distribution/logistic" factories. We just need a way to not clog up the picker. 

> The challenge then, is that this 'distribution' factory would now show in other factories as a valid import. Again, the example of the computer, if we put plastic here, the computer factory will draw from it again.

Why is this a challenge? Perhaps I am misunderstanding your intention. Supporting the ability to import plastic from various sources into a "distribution" center, which then can act as a valid "factory" for Computer factories to import from, seems like the capability we are suggesting. Unless of course the challenge relates to your next point (storage reserve) and not the original point (a good implementation for unrestricted import to factories).

> Therefore, I think we should also consider the storage aspect of this - where you can mark a particular production line as going to 'storage' and hence, is not available as a surplus. Now the excess part would not be available for export. 

If you are looking for some kind of reminder indicating you wish to reserve a certain portion of a factory's production, perhaps a "Reserve" rate could be implemented in factory outputs. I suppose the ability to define a reserve production rate could help prevent players from accidentally tapping into production originally intended to be reserved (as in collecting elevator parts or building supplies, etc). But again, I don't see why this requires a special "Storage" factory implementation. As a player, i would expect that all factories could be set to reserve a portion of its production (if that is a feature you all want to the planner to implement), not just factories that are ticked as "storage". TBH, if additional feature were restricted to "storage" factories (the ability to freely import or set reserve production) I would just tick the storage tick-box for all factories.

**Comment by Maelstromeous** [comment]
@450Volts @samsmithnz 

TL;DR: I'm open to having a factory be flaggable as "distribution center" where the restriction of "imports must go into production, production must go to exports" is unlocked to remove the production step (effectively turning all imports into exportable parts), but beyond that I don't see a means to handle this effectively within the planner.

This is an implementation that is quite easily achievable without much more than a day or two's work, because of the changes I've made in Update 3, where part calculations and export candidates are now centralized, the logic can be easily adjusted and tested for.

> Unifying the ability to freely import parts to any factory avoids the need to set separate rules for "normal" factories and "Distribution/logistic" factories. We just need a way to not clog up the picker.

I'm thinking the opposite, simply have a toggle on the factory "unlocking" the ability to import from any factory. The enhancements I've made to Imports in update 3 should make this a bit less combersome than it was previously, at least when you exhaust a factory's parts it drops off the list.

> Supporting the ability to import plastic from various sources into a "distribution" center, which then can act as a valid "factory" for Computer factories to import from, seems like the capability we are suggesting.

This is **technically** possible, but we will have to again have a form of toggle to "unlock" the ability for the factory to have export candidates directly sourced from imports. Right now the logic is as such:

1. Importing into a factory is exclusively used only by products. They are not marked for export.
2. Products which **are not raw** are marked for export. This includes byproducts.
3. Exportable products are then avaialble for factories which scan for imports of their part requirements.

We would have to have a special type of factory to remove the "production" layer and effectively re-ship the imports as exports. I will not be adding this as default behaviour where imports can from from literally any factory with any part, that would be madness, **but** I am willng to enable specially flagged factories to be able to do this. This works nicely as this means only one extra factory is added to the import candidate list and with the recent import changes it drops off the parts of said distribution "factory" as they are selected.\

> So if you are looking for some kind of reminder indicating you wish to reserve a certain portion of a factory's production, maybe a "Reserve" rate could be implemented in factory outputs.

I see no widespread reason to implement this. It's too niche. Simply overproduce and under-demand, there's your reserve.

The only exception to this may be Consumer Sinks we're discussing where a certain portion of a factory's production is dedicated to being sunk or a combination of dimensional storage then sunk (ultimately being sinked). Beyond this though, I don't see any reason to dedicate a portion of production for "storage", storage is finite, and will eventually clog unless the excess is sunk, to which to me defeats the whole purpose of defining a factory as storage as, quite simply, it's just a very very large buffer that _eventually_ goes somewhere.

I will be very clear on this, we are not supporting any means of tracking storage levels or any form of storage facility in that manner in the planner, there is simply no point to it.

**Comment by samsmithnz** [comment]
Sounds reasonable to me. What about parts that need to be tagged as in storage? This isn't tracking the storage- just marking the parts as used. The main use case for this slightly over producing for storage and not having the planner trim that excess.

**Comment by Maelstromeous** [comment]
So you mean:

- Fac 1 produces 10 iron plates, ships to Distribution
- Distribution recieves 10 iron plates, exports 8 plates to Fac 2
- Fac 2 receives 8 plates from distribution

That the 2 excess plates would require trimming or are we going to assume they are sunk? As that would cause a backlog.

**Comment by samsmithnz** [comment]
If we want to call it sinking instead of/as well as storage, that works for me- we are talking about the same thing I believe. The challenge is that a part of a production needs to be allocated as "allocated surplus". What if
- There is a flag on a factory that makes it 'distribution factory'. This flag enables imports to show all factories with surplus.
- After a factory is marked as 'distribution', there is another flag that marks it as 'sink/storage'. This flag marks all imported products into the factory as allocated (no longer surplus). These products can't be exported.

Like the overclocking (and sloops), I don't think we need to get to the level of each machine - this should be up to the user. We just need to make sure every part is 'allocated' in the plan, whether it's imported in another production line, (via distribution or not), sunk, or put in storage. I don't think it matters if the storage can back up - again - that is a game problem the user can manage themselves. We aren't fool-proofing the factory with smart splitters (error handlers); we are just showing the factory architecture...

**Comment by immerrr** [comment]
i'd like to leave a suggestion for a feature that sounds close to what was discussed here, setting an "allocated surplus" for a part.

for example, i have a factory that produces wire/cable for all the other factories, but also sets aside N/min items of each to use when building electricity stuff. the surplus is usually sent into a belt that ends with "smartsplitter(overflow to sink) -> ind.storage -> dim.storage", so it doesn't back up the main production line, but i have some wire and cable at hand at all times.

it would be nice to be able to say "this factory should have X items of surplus at all times", so that one doesn't need to manually add those X items every time they do "fix production" automatically. 

some ideas on the requirements:
- set surplus allocation as "minimum X/min", with the wording that conveys that going over that limit is not something that needs to be fixed (but it can still be trimmed down up to the limit as usual)
- it should only work for solid objects, since liquids and gases can't be sunk, and for packaging one needs to account for the container production
- the configuration should probably live in the "Satisfaction" column, as this is where the surplus is shown right now, maybe show up the configuration form as a modal on click, or as part of "satisfaction breakdowns"
![image](https://github.com/user-attachments/assets/3eac13ef-2c66-4a65-b11a-5f3507d85c6f)
- there is a question though on how to visually show that the allocated surplus requirement is not fulfilled: the row should turn red if the requirement is not met, but it may not be clear if the row is red, but current surplus is positive, so maybe some extra clarification is needed in this case
![image](https://github.com/user-attachments/assets/74113451-f170-49d6-afd1-d4bcb6208092)
- it is also unclear to me if there should be a way to create an "allocated surplus" just from the "Satisfaction" section; on one hand, it is not strictly needed as an item can be added as a product manually first; on the other hand, creating a way to "allocate a surplus" without producing an item would open the door to logistic/storage centres described in this issue: one would configure the expected surplus (0 for pure logistic, and X > 0 for storage ones) and then these surplus configs could be used to populate import/export dropdowns

**Comment by samsmithnz** [comment]
Possible solution.
1. Add a new section, under imports, labeled "Storage/Sink"
2. A similar list to imports, where you can add items from the internal list that have a surplus. 
3. In the quantity value, when you enter a value, that quantity is allocated and therefore doesn't show as a surplus anymore. 
4. Whether the part is sunk or stored I think doesn't matter - it's just allocating the surplus. It's up to the user to figure out this part, although we could just add a dropdown to show either "stored" or "sunk"

**Comment by Maelstromeous** [comment]
Had further thoughts on this.

For Distribution Centers, this is reletively simple to implement:

1. Add a toggle on the factory to enable "Distribution mode"
2. "Distribution mode" disables the ability to add Products to a factory, and disables the checks against products.
3. The user imports things from **producing** factories like they would.
4. The user then has **exports** available to **consumer** factories.

The new `part` data accounts for imports and exports now, so this should be thankfully quite a simple change, it's just a bunch of UI tweaks.

**Comment by samsmithnz** [comment]
I think if the toggle switch also has "storage/sink", and doesn't allow exports, it would also solve that issue, wouldn't it?

**Comment by Maelstromeous** [comment]
Possibly. I'll need to map it out in my mind more.
